### ISSUE TITLE
Wrong behavior using dynamic task included in role dependency

### ISSUE THREAD

**Issue by Erouan50** [ansible/ansible#29052]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
core

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 02a362e7de) last updated 2017/09/05 18:14:15 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'~/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = ~/Documents/ansible/lib/ansible
  executable location = ~/Documents/ansible/bin/ansible
  python version = 2.7.13 (default, Jun 26 2017, 10:20:05) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```

##### CONFIGURATION
Defaults

##### OS / ENVIRONMENT
N/A

##### SUMMARY
Variable used in dynamic task included via a role dependency doesn't change when the parent role is used several times in the same play with a different value for the variable. The first time, the variable value is correct. After the value will be still the same during the play execution. Usually, it should change. With a static include, it does work.

##### STEPS TO REPRODUCE
Example:
- Two roles, A and B
- B depends on A
- A has a dynamic task included which displays the *name* variable
- The playbook calls the role B several times with different value for the name variable

Here are my files to reproduce the issue:

````
.
├── roles
│   ├── A
│   │   ├── meta
│   │   │   └── main.yml
│   │   └── tasks
│   │       ├── main.yml
│   │       └── my_include.yml
│   └── B
│       └── meta
│           └── main.yml
└── test.yml
````

*roles/A/meta/main.yml*
````
allow_duplicates: yes
````
*roles/A/tasks/main.yml*
````
- debug: var=name
- include_tasks: "my_include.yml"
````
*roles/A/tasks/my_include.yml*
````
- debug: var=name
````
*roles/B/meta/main.yml*
````
dependencies:
  - role: A
````
*test.yml*
````
- hosts: localhost
  connection: local
  roles:
    - role: B
      name: test_1
    - role: B
      name: test_2
````

##### EXPECTED RESULTS
````
PLAY [localhost] ******************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_1"
}

TASK [A : include_tasks] **********************************************************************************************************************************************************************************************************************
included: ~/Documents/test/roles/A/tasks/my_include.yml for localhost

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_1"
}

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_2"
}

TASK [A : include_tasks] **********************************************************************************************************************************************************************************************************************
included: ~/Documents/test/roles/A/tasks/my_include.yml for localhost

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_2"
}

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0 
````
The last debug should be "test_2"

##### ACTUAL RESULTS
The actual value of the last debug is "test_1"

````
PLAY [localhost] ******************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_1"
}

TASK [A : include_tasks] **********************************************************************************************************************************************************************************************************************
included: ~/Documents/test/roles/A/tasks/my_include.yml for localhost

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_1"
}

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_2"
}

TASK [A : include_tasks] **********************************************************************************************************************************************************************************************************************
included: ~/Documents/test/roles/A/tasks/my_include.yml for localhost

TASK [A : debug] ******************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "name": "test_1"
}

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0   

````

**Comment by samdoran** [comment]
cc @bcoca

**Comment by Sodki** [comment]
I'm hitting the same issue. The only difference in the skeleton is that on B I'm using `include_role` instead of using `meta`. But the result is the same. This unfortunately means that if you've been doing the right thing and reusing roles, you end up with a broken system and there's no workaround.

**Comment by sivel** [comment]
This issue was resolved in f1c458303a1ad00675497597e1f7fbc0539e314e and included in the 2.4.3 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
